File: blockchain/src/main/java/pt/um/lei/masb/blockchain/utils/SHA256Encrypter.java
package pt.um.lei.masb.blockchain.utils;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.UnsupportedEncodingException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public final class SHA256Encrypter implements Crypter {
    private final static Logger LOGGER = LoggerFactory.getLogger(SHA256Encrypter.class);

    //Applies Sha256 to a string and returns the result.
    public String applyHash(String input) {
        try {
            var digest = MessageDigest.getInstance("SHA-256");
            //Applies sha256 to our input,
            var hash = digest.digest(input.getBytes("UTF-8"));
            var hexString = new StringBuilder(); // This will contain the hash as hexadecimal
            for (byte aHash : hash) {
                String hex = Integer.toHexString(0xff & aHash);
                if (hex.length() == 1) {
                    hexString.append('0');
                }
                hexString.append(hex);
            }
            return hexString.toString();
        } catch (NoSuchAlgorithmException | UnsupportedEncodingException e) {
            LOGGER.error("", e.getMessage());
            throw new RuntimeException("Apply SHA256 problem", e);
        }
    }

    @Override
    public long hashSize() {
        return 32;
    }
}

File: blockchain/src/main/kotlin/pt/um/lei/masb/blockchain/utils/SHA256Encrypter.kt
package pt.um.lei.masb.blockchain.utils

import mu.KLogging
import pt.um.lei.masb.blockchain.Hash
import java.security.MessageDigest

class SHA256Encrypter : Crypter {
    companion object : KLogging()

    val digester = MessageDigest.getInstance(
        "SHA-256"
    )

    /**
     * Applies Sha256 to a string [input] and returns the resulting [Hash].
     */
    override fun applyHash(input: String): Hash =
        digester.digest(input.toByteArray())

    /**
     * Applies Sha256 to a string [input] and returns the resulting [Hash].
     */
    override fun applyHash(input: ByteArray): Hash =
        digester.digest(input)

    override val hashSize: Long = 32
}
